src/serde_helpers.rs

```diff
@@ -0,0 +1,175 @@
+) -> Result<DateTime<Utc>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bson = Bson::deserialize(deserializer)?;
+    match bson {
+        Bson::DateTime(date) => Ok(date),
+        _ => Err(de::Error::custom(
+            "cannot convert extended JSON to DateTime",
+        )),
+    }
+}
+
+/// Deserializes a DateTime from the extended JSON representation.
+pub fn deserialize_bson_datetime_from_ext_json<'de, D>(
+    deserializer: D,
+) -> Result<crate::DateTime, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bson = Bson::deserialize(deserializer)?;
+    match bson {
+        Bson::DateTime(date) => Ok(crate::DateTime(date)),
+        _ => Err(de::Error::custom(
+            "cannot convert extended JSON to DateTime",
+        )),
+    }
+}
+
+/// Deserializes an ISO string from a DateTime.
+pub fn deserialize_iso_string_from_datetime<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let date = crate::DateTime::deserialize(deserializer)?;
+    Ok(date.to_string())
+}
+
+/// Serializes an ISO string as a DateTime.
+pub fn serialize_iso_string_as_datetime<S: Serializer>(
+    val: &str,
+    serializer: S,
+) -> Result<S::Ok, S::Error> {
+    let date = DateTime::from_str(val)
+        .map_err(|_| ser::Error::custom(format!("cannot convert {} to DateTime", val)))?;
+    Bson::DateTime(date).serialize(serializer)
+}
+
+/// Deserializes an ObjectId from the extended JSON representation.
+pub fn deserialize_object_id_from_ext_json<'de, D>(deserializer: D) -> Result<ObjectId, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bson = Bson::deserialize(deserializer)?;
+    match bson {
+        Bson::ObjectId(oid) => Ok(oid),
+        _ => Err(de::Error::custom(
+            "cannot convert extended JSON to ObjectId",
+        )),
+    }
+}
+
+/// Serializes a hex string as an ObjectId.
+pub fn serialize_hex_string_as_object_id<S: Serializer>(
+    val: &str,
+    serializer: S,
+) -> Result<S::Ok, S::Error> {
+    match ObjectId::with_string(val) {
+        Ok(oid) => oid.serialize(serializer),
+        Err(_) => Err(ser::Error::custom(format!(
+            "cannot convert {} to ObjectId",
+            val
+        ))),
+    }
+}
+
+/// Serializes a Uuid as a Binary.
+pub fn serialize_uuid_as_binary<S: Serializer>(
+    val: &Uuid,
+    serializer: S,
+) -> Result<S::Ok, S::Error> {
+    let binary = Binary {
+        subtype: BinarySubtype::Uuid,
+        bytes: val.as_bytes().to_vec(),
+    };
+    binary.serialize(serializer)
+}
+
+/// Deserializes a Uuid from a Binary.
+pub fn deserialize_uuid_from_binary<'de, D>(deserializer: D) -> Result<Uuid, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let binary = Binary::deserialize(deserializer)?;
+    if binary.subtype == BinarySubtype::Uuid {
+        if binary.bytes.len() == 16 {
+            let mut bytes = [0u8; 16];
+            for (i, byte) in bytes.iter_mut().enumerate() {
+                *byte = binary.bytes[i];
```

Comment:
There is a `TryFrom` implementation that makes this conversion a lot simpler, but it's only available in 1.48+. Since we support lower compiler versions this roundabout way is necessary -- the index is safe since the size of the vector is guaranteed in the previous check.

Reply:
I think it might be more idiomatic to use [`copy_from_slice`](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) here rather than copying each byte manually. Clippy sometimes warns about things like this, although I'm not sure if it can detect it all of the time.